### PR TITLE
Add support for date/time/datetime fields

### LINKED ISSUE #585: Missing support for date/time types?
The library currently supports the BigQuery timestamp type, but apparently not the other date/time types (date, datetime, time, see https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types).

@jskeet I can submit a PR to add support for these, but I remembered you're the Nodetime guy so I thought I'd check with you first:)

### PR DESCRIPTION
**Pull request by jskeet**
Next we'll want to elide BigQueryDbType and BigQueryParameterType...

Fixes #585.

// cc @roji

### CODE REVIEW

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryDbType.cs [hunk 1]
```diff
@@ -36,6 +36,18 @@ public enum BigQueryDbType
         /// </summary>
         Timestamp,
         /// <summary>
+        /// A civil date in the Gregorian calendar.
+        /// </summary>
+        Date,
+        /// <summary>
+        /// A civil time-of-day with up to microsecond precision.
```
Comment (chrisdunelm) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryDbType.cs]: "Up to microsecond precision" is a bit ambiguous, that could mean that it only supports hour precision. What's the minimum guaranteed precision?

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryRow.cs [hunk 2]
```diff
@@ -44,11 +44,19 @@ internal BigQueryRow(TableRow rawRow, TableSchema schema)
         private static readonly Func<string, string> StringConverter = v => v;
         private static readonly Func<string, long> Int64Converter = v => long.Parse(v, CultureInfo.InvariantCulture);
         private static readonly Func<string, double> DoubleConverter = v => double.Parse(v, CultureInfo.InvariantCulture);
-        private static readonly Func<string, DateTime> DateTimeConverter = v => new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc).AddSeconds(DoubleConverter(v));
+        // AddSeconds rounds to the nearest millisecond, for some reason.
+        private static readonly Func<string, DateTime> TimestampConverter = v =>
+        {
+            double seconds = DoubleConverter(v);
+            long microseconds = (long) (seconds * 1000000);
```
Comment (chrisdunelm) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryRow.cs]: 1e6 might be clearer?

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryRow.cs [hunk 3]
```diff
@@ -44,11 +44,19 @@ internal BigQueryRow(TableRow rawRow, TableSchema schema)
...
+            long ticks = microseconds * 10;
+            return new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc).AddTicks(ticks);
+        };
```
Comment (chrisdunelm) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryRow.cs]: I don't really understand this change - it appears to be doing the same as before? Or is it more precise somehow?

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/QueryTest.cs [hunk 4]
```diff
@@ -196,5 +199,105 @@ public void EmptyQueryResults_GetResultSet()
...
+
+            Assert.Equal("Hello", row["String"]);
+            Assert.Equal(Encoding.UTF8.GetBytes("Hello"), row["Bytes"]);
+            Assert.Equal(true, row["Boolean"]);
+            Assert.Equal(1234567890L, row["Int64"]);
+            Assert.Equal(1.234567, row["Float64"]);
+            Assert.Equal(new DateTime(2016, 11, 29, 0, 0, 0, DateTimeKind.Unspecified), row["Date"]);
+            // There are 10 ticks per microsecond.
+            Assert.Equal(new TimeSpan(0, 12, 34, 56, 123) + TimeSpan.FromTicks(4560), row["Time"]);
+            Assert.Equal(new DateTime(2016, 11, 29, 12, 34, 56, 123) + TimeSpan.FromTicks(4560), row["DateTime"]);
```
Comment (chrisdunelm) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/QueryTest.cs]: Specify a kind for clarity?

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryDbType.cs (hunk 1 above)
Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryDbType.cs]: I think that *used* to be the official wording in the docs, but it's now just "with microsecond precision", so will match that.

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryRow.cs (hunk 2 above)
Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryRow.cs]: Done.

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryRow.cs (hunk 3 above)
Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryRow.cs]: Yes - if you call `AddSeconds`, that rounds the argument to 3dps (millisecond precision) for no obvious reason. We're just converting the seconds into ticks ourselves, so we can call AddTicks that doesn't perform any rounding.

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/QueryTest.cs (hunk 4 above)
Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/QueryTest.cs]: Done.

**Pull request merged**